### PR TITLE
gh-54781: Move Lib/tkinter/test/test_ttk/ to Lib/test/test_ttk/

### PR DESCRIPTION
**Pull request by vstinner**
Move Lib/tkinter/test/ to Lib/test/test_tkinter/.

Rename test_tk to test_tkinter, and rename test_ttk_guionly to test_ttk.

* gh-54781

### CODE REVIEW

**Comment by vstinner** [comment]
I can still run tests on an newly installed (patched) Python 3.12:

```
$ /opt/py3.12/bin/python3.12 -m test test_tkinter -u all -v
(...)
Ran 731 tests in 6.853s

$ /opt/py3.12/bin/python3.12 -m test test_ttk -u all -v
(...)
Ran 308 tests in 7.905s

$ /opt/py3.12/bin/python3.12 -m test test_tkinter
test_tkinter skipped -- Use of the 'gui' resource not enabled
test_tkinter skipped (resource denied)

$ /opt/py3.12/bin/python3.12 -m test test_ttk
test_ttk skipped -- Use of the 'gui' resource not enabled
test_ttk skipped (resource denied)
```

**Comment by vstinner** [comment]
> Rename test_tk to test_tkinter

For me, it's disturbing when tests don't have the same name than the tested module. asyncio => test_asyncio. tkinter => test_tk ???

> rename test_ttk_guionly to test_ttk

I can keep test_ttk_guionly name if you prefer, but for me, tests should use the GUI. Only the special tests test_ttk_textonly should have a special name, no?

**Comment by vstinner** [comment]
cc @terryjreedy @serhiy-storchaka

**Comment by vstinner** [comment]
This PR is special. I made the new directory tree "flat". ``support.py`` and ``test_simpledialog.py`` now lives in ``Lib/test/test_tkinter/``. Previously, ``test_simpledialog.py`` was in a ``test_tkinter/`` sub-directory.

Before:

```
Lib/tkinter/test/
Lib/tkinter/test/README
Lib/tkinter/test/__init__.py
Lib/tkinter/test/support.py
Lib/tkinter/test/test_tkinter
Lib/tkinter/test/test_tkinter/__init__.py
Lib/tkinter/test/test_tkinter/test_colorchooser.py
Lib/tkinter/test/test_tkinter/test_font.py
Lib/tkinter/test/test_tkinter/test_geometry_managers.py
Lib/tkinter/test/test_tkinter/test_images.py
Lib/tkinter/test/test_tkinter/test_loadtk.py
Lib/tkinter/test/test_tkinter/test_messagebox.py
Lib/tkinter/test/test_tkinter/test_misc.py
Lib/tkinter/test/test_tkinter/test_simpledialog.py
Lib/tkinter/test/test_tkinter/test_text.py
Lib/tkinter/test/test_tkinter/test_variables.py
Lib/tkinter/test/test_tkinter/test_widgets.py
Lib/tkinter/test/test_ttk
Lib/tkinter/test/test_ttk/__init__.py
Lib/tkinter/test/test_ttk/test_extensions.py
Lib/tkinter/test/test_ttk/test_style.py
Lib/tkinter/test/test_ttk/test_widgets.py
Lib/tkinter/test/widget_tests.py
```

After:

```
Lib/test/test_ttk/
Lib/test/test_ttk/__init__.py
Lib/test/test_ttk/test_extensions.py
Lib/test/test_ttk/test_style.py
Lib/test/test_ttk/test_widgets.py

Lib/test/test_tkinter/
Lib/test/test_tkinter/README
Lib/test/test_tkinter/__init__.py
Lib/test/test_tkinter/support.py
Lib/test/test_tkinter/test_colorchooser.py
Lib/test/test_tkinter/test_font.py
Lib/test/test_tkinter/test_geometry_managers.py
Lib/test/test_tkinter/test_images.py
Lib/test/test_tkinter/test_loadtk.py
Lib/test/test_tkinter/test_messagebox.py
Lib/test/test_tkinter/test_misc.py
Lib/test/test_tkinter/test_simpledialog.py
Lib/test/test_tkinter/test_text.py
Lib/test/test_tkinter/test_variables.py
Lib/test/test_tkinter/test_widgets.py
Lib/test/test_tkinter/widget_tests.py
```

**Comment by vstinner** [comment]
Windows (x64) job failed but I cannot get the logs. I cancelled the workflow. I still cannot see the logs. Strange.

**Comment by zware** [comment]
> Windows (x64) job failed but I cannot get the logs. I cancelled the workflow. I still cannot see the logs. Strange.

The [failure](https://github.com/python/cpython/runs/6988077489?check_suite_focus=true) is unrelated, see #94068.

File: Lib/test/test_tkinter/__init__.py [hunk 1]
```diff
@@ -12,9 +12,6 @@
 # Skip test if tk cannot be initialized.
 support.requires('gui')
 
-def load_tests(loader, tests, pattern):
-    return loader.discover('tkinter.test.test_tkinter')
-
 
-if __name__ == '__main__':
```
Comment (serhiy-storchaka) [inline Lib/test/test_tkinter/__init__.py]: Does `python -m test.test_tkinter` still work?

File: Lib/test/test_tkinter/__init__.py [hunk 2]
```diff
@@ -12,9 +12,6 @@
...
-    unittest.main()
+def load_tests(*args):
+    return support.load_package_tests(os.path.dirname(__file__), *args)
```
Comment (serhiy-storchaka) [inline Lib/test/test_tkinter/__init__.py:18]: Why not use just `loader.discover()`?

**Comment by serhiy-storchaka** [comment]
Should it be backported?

**Comment by terryjreedy** [comment]
I would prefer that ttk tests be included under test_tkinter, with names like `test_ttk_widgets`, so that `python -m test.test_tkinter` (and `python -m test -ugui test_tkinter`) tests everything.  But re-organization details are up to Serhiy.

**Comment by terryjreedy** [comment]
This patch will require changes to the scripts *nix distributions use to make tkinter (and IDLE) a separate install.  I don't know how they would feel about backports.

File: Lib/test/test_tkinter/__init__.py (hunk 1 above)
Comment (merwok) [reply Lib/test/test_tkinter/__init__.py]: It would work if a `__main__` module gets added.  But it was `test.test_tk` before, not `test.test_tkinter`.

Comment (vstinner) [reply Lib/test/test_tkinter/__init__.py]: Ah you care about this command? If you do, I can add a ``__main__.py`` file.

File: Lib/test/test_tkinter/__init__.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_tkinter/__init__.py:18]: I copy/paste the code from test_asyncio, it works, so I didn't change it :-)

**Comment by vstinner** [comment]
> Should it be backported?

This specific PR change the name of two tests, so no, it should not be backported.

It also changes the build system (Linux, VS project). Honestly, it sounds risky and I would prefer to not backport it.

**Comment by vstinner** [comment]
> I would prefer that ttk tests be included under test_tkinter

Oh. I tried to minimize changes in this PR. But I'm open to reorganizing tests.

@serhiy-storchaka: Are you fine with the 2 proposed tests, or do you want to merge them?

File: Lib/test/test_tkinter/__init__.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/test/test_tkinter/__init__.py]: Yes, please add it. And test different ways of running tests:
```
python -m test.test_xxx
python -m unittest test.test_xxx
python -m test test_xxx
```

File: Lib/test/test_tkinter/__init__.py (hunk 2 above)
Comment (serhiy-storchaka) [reply Lib/test/test_tkinter/__init__.py:18]: Does `loader.discover()` work with zipimport?

**Comment by serhiy-storchaka** [comment]
> Are you fine with the 2 proposed tests, or do you want to merge them?

I was going to ask about merging them, but I am fine with the current state of this PR. If you are going to merge `test_ttk` into `test_tkinter`, do not forget about `test_tcl`.

Maybe do this in a separate PR?

File: Lib/test/test_tkinter/__init__.py (hunk 2 above)
Comment (vstinner) [reply Lib/test/test_tkinter/__init__.py:18]: > Does loader.discover() work with zipimport?

I have no idea. If there is an issue, I suggest to fix all load_tests() functions of Lib/test/ in a separated PR.

File: Lib/test/test_tkinter/__init__.py (hunk 1 above)
Comment (vstinner) [reply Lib/test/test_tkinter/__init__.py]: I added ``__main__`` sub-modules. Example:

```
./python -m test.test_ttk -v
./python -m test test_ttk -v -u all
./python -m unittest test.test_ttk -v
```

The 3 commands end with: ``Ran 308 tests``.

**Comment by vstinner** [comment]
> I was going to ask about merging them, but I am fine with the current state of this PR. If you are going to merge test_ttk into test_tkinter, do not forget about test_tcl.

Oh, merging 3 Tkinter tests sound out of my skills. I don't know this module and I would prefer if someone else do it.

Here I'm trying to focus on fixing the 12 years old issue #54781 :-) Perfect is the enemy of good ;-)

**Comment by vstinner** [comment]
@serhiy-storchaka: Does it look good to you like that?

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

But I see that Git does not recognize renaming of test_tk.py. It breaks history and future backports. Can something be done with this?

**Comment by vstinner** [comment]
> But I see that Git does not recognize renaming of test_tk.py

I didn't rename test_tk.py to test_tkinter/__init__.py. I removed test_tk.py and copied test_asyncio/__init__.py to test_tkinter/__init__.py, and then modified test_tkinter/__init__.py.

test_tk.py was just 18 lines of code, I don't think its file history was very important.

**Comment by vstinner** [comment]
On the other side, ``git blame Lib/test/test_tkinter/test_widgets.py`` and ``git log --follow Lib/test/test_tkinter/test_widgets.py`` kept the file history for example.

**Pull request merged**